### DISCUSSION TITLE
usb midi on STM32L151

### DISCUSSION THREAD

**Discussion by ctag-fh-kiel** [hathach/tinyusb#994] [Q&A] [answered]
Hi, I have been trying to get tinyusb with usb midi running on an STM32L151C8.

No success :( 

My config file is here: 
https://github.com/ctag-fh-kiel/stm32l151-tusb-midi/blob/master/Core/Inc/tusb_config.h

I extended the driver files:
https://github.com/ctag-fh-kiel/tinyusb/tree/master/src/portable/st/stm32_fsdev
Changes include: Correct interrupt en/disable routines (LP USB Interrupt), PMA_LENGTH (512u)

I am calling tusb_init(); and tud_task(); correctly
Also i call tud_int_handler(0); in the USB_LP_IRQHandler(void) ISR

I have tried __HAL_RCC_USB_CLK_ENABLE(); before and after tusb_init(); 
The USB FS device is enabled with CubeMX. 48MHz derived from PLL which is fed an external 12MHz signal.

If I use the STM32 middleware, I can get an HID device running, so electrically everything seems ok.

Anyone has an idea? If I get it working on the device, I'd like to create a pull request for the drivers, maybe it would help someone as well...

My project repo is: https://github.com/ctag-fh-kiel/stm32l151-tusb-midi
My Tiny USB fork is: https://github.com/ctag-fh-kiel/tinyusb

Thanks for any advice :)

**Comment by hathach** [comment]
L1 haven't tested with TinyUSB just yet, even those STM32 use the same controller, they are often slightly different e.g L0 work, but L4 does not work with TinyUSB yet. I would suggest you try to add new BSP to repo and try to run stock example e.g cdc_msc before using it with other platform such as cube to limit the variable factor.

**Comment by ctag-fh-kiel** [comment, marked as answer]
It seems I have gotten it to work, what's needed is to manually enable the internal D+ pull-up with __HAL_SYSCFG_USBPULLUP_ENABLE(); after HAL USB config.

Where could the be placed ideally in tinyusb? I would create a merge request once tested a bit... 

Cheers :)

**Reply by hathach** [reply]
if it is pull up, it is probably be here, you may to do #ifelse based on the MCU or present of register to make it more common acorss ST MCU  https://github.com/hathach/tinyusb/blob/master/src/portable/st/stm32_fsdev/dcd_stm32_fsdev.c#L273

**Reply by ctag-fh-kiel** [reply]
These are the lines which would be needed without the macro, it seems some definitions are different in the file you referred?
SET_BIT(SYSCFG->PMC, SYSCFG_PMC_USB_PU);
CLEAR_BIT(SYSCFG->PMC, SYSCFG_PMC_USB_PU);

Any advice how this matches to:

#if defined(USB_BCDR_DPPU)

// Disable internal D+ PU
void dcd_disconnect(uint8_t rhport)
{
  (void) rhport;
  USB->BCDR &= ~(USB_BCDR_DPPU);
}

// Enable internal D+ PU
void dcd_connect(uint8_t rhport)
{
  (void) rhport;
  USB->BCDR |= USB_BCDR_DPPU;
}

#endif

**Reply by hathach** [reply]
you could do something similar to existing code, here is what I propose. If it works with your board, please consider to make an PR to get it into the upstream :) 

```C

#if defined(USB_BCDR_DPPU)

// Disable internal D+ PU
void dcd_disconnect(uint8_t rhport)
{
  (void) rhport;
  USB->BCDR &= ~(USB_BCDR_DPPU);
}

// Enable internal D+ PU
void dcd_connect(uint8_t rhport)
{
  (void) rhport;
  USB->BCDR |= USB_BCDR_DPPU;
}

#elif defined(SYSCFG_PMC_USB_PU)

void dcd_disconnect(uint8_t rhport)
{
  (void) rhport;
  SYSCFG->PMC &= ~(SYSCFG_PMC_USB_PU);
}

// Enable internal D+ PU
void dcd_connect(uint8_t rhport)
{
  (void) rhport;
  SYSCFG->PMC |= SYSCFG_PMC_USB_PU;
}

#endif

**Reply by ctag-fh-kiel** [reply]
just submitted the PR

**Reply by hathach** [reply]
thank you, reviewed and merged !!! :tada:
